go/exorelay/index.go

```diff
@@ -0,0 +1,143 @@
+		closed:          false,
+		exoSocket:       exosocket.NewExoSocket(config),
+		replyChannels:   ReplyChannelMapping{},
+		messageHandlers: MessageHandlerMapping{},
+	}
+}
+
+// Connect connects to ExoCom and starts listening for messages
+func (e *ExoRelay) Connect() error {
+	err := e.exoSocket.Connect()
+	if err != nil {
+		return err
+	}
+	go e.listenForMessages()
+	return nil
+}
+
+// Close disconnects from ExoCom
+func (e *ExoRelay) Close() error {
+	if e.closed {
+		return nil
+	}
+	err := e.exoSocket.Close()
+	e.closed = true
+	return err
+}
+
+// RegisterHandler adds a handler for the given message name
+func (e *ExoRelay) RegisterHandler(messageName string, handler MessageHandler) {
+	e.messageHandlersMutex.Lock()
+	e.messageHandlers[messageName] = handler
+	e.messageHandlersMutex.Unlock()
+}
+
+// Send sends a message
+func (e *ExoRelay) Send(messageOptions exosocket.MessageOptions) (*structs.Message, error) {
+	return e.exoSocket.Send(messageOptions)
+}
+
+// WaitForActivity returns the first message received with the given activityID
+func (e *ExoRelay) WaitForActivity(activityID string, timeout time.Duration) (*structs.Message, error) {
+	messageChannel := make(chan *structs.Message)
+	e.addReplyChannel(activityID, messageChannel)
+	defer e.removeReplyChannel(activityID)
+	select {
+	case message := <-messageChannel:
+		return message, nil
+	case <-time.After(timeout):
+		return nil, fmt.Errorf("Did not receive a reply for '%s' after %s", activityID, timeout.String())
+	}
+}
+
+// Helpers
+
+func (e *ExoRelay) addReplyChannel(activityID string, messageChannel chan *structs.Message) {
+	e.replyChannelsMutex.Lock()
+	e.replyChannels[activityID] = messageChannel
+	e.replyChannelsMutex.Unlock()
+}
+
+func (e *ExoRelay) buildRequest(message structs.Message) Request {
+	return Request{
+		Message: message,
+		Reply: func(name string, payload structs.MessagePayload) error {
+			_, err := e.exoSocket.Send(exosocket.MessageOptions{
+				Name:       name,
+				Payload:    payload,
+				ActivityID: message.ActivityID,
+				Auth:       message.Auth,
+				IsSecurity: message.IsSecurity,
+			})
+			return err
+		},
+		Send: func(name string, payload structs.MessagePayload) (*structs.Message, error) {
+			return e.exoSocket.Send(exosocket.MessageOptions{
+				Name:    name,
+				Payload: payload,
+				Auth:    message.Auth,
+			})
+		},
+		WaitForActivity: e.WaitForActivity,
+	}
+}
+
+func (e *ExoRelay) listenForMessages() {
+	messageChannel := e.exoSocket.GetMessageChannel()
+	for {
+		message, ok := <-messageChannel
+		if !ok {
+			return
+		}
+		go e.sendToMessageHandler(message)
+		go e.sendToReplyChannel(message)
+	}
+}
+
+func (e *ExoRelay) removeReplyChannel(activityID string) {
+	e.replyChannelsMutex.Lock()
+	delete(e.replyChannels, activityID)
+	e.replyChannelsMutex.Unlock()
+}
+
+func (e *ExoRelay) sendToMessageHandler(message structs.Message) {
+	e.messageHandlersMutex.RLock()
+	defer e.messageHandlersMutex.RUnlock()
```

Comment:
This keeps the `messageHandlersMutex` open for the entirety of building the request, thereby limiting concurrency. Could you unlock the mutex already at line 131?